File: phenotyping/scripts/RNA_editing/shared_samples_sites_matrix.pl
#!/usr/bin/env perl
# Perl script for making edit level matrix
# Originally from https://github.com/vargasliqin/GTEx_edQTL

use strict;
use warnings;
use Getopt::Long;
use IO::Uncompress::Gunzip qw(gunzip $GunzipError);
use Cwd;

my ($path_edit_files, $samples_file, $output_file, $mincov, $minsamps);

# Parsing command-line options
GetOptions(
    'path_to_edit_files=s'    => \$path_edit_files,
    'samples_file=s'          => \$samples_file,
    'output_file=s'           => \$output_file,
    'min_coverage=i'          => \$mincov,
    'min_samples=i'           => \$minsamps,
) or die "Error in command line arguments: $!\n";

# Adding "/" to thr end of path if it's not there already
$path_edit_files .= "/" unless $path_edit_files =~ /\/$/;

print "Path to Edit Files: $path_edit_files\n";
print "Samples File: $samples_file\n";
print "Output File: $output_file\n";
print "Min Coverage: $mincov\n";
print "Min Samples: $minsamps\n";

# Read samples from file
my @samples;
open(my $SAMPLES, "<", $samples_file) or die "Could not open samples file '$samples_file': $!\n";
while(<$SAMPLES>) {
    chomp;
    push @samples, $_;
}
close $SAMPLES;

my %sitehash;
my %totalhash;
my %lvlhash;

foreach my $sample (@samples) {
    my $file = $path_edit_files . $sample . ".rnaeditlevel.tsv.gz";
    unless (-e $file) {
        print "Warning: File not found for sample $sample: $file\n";
        next;
    }
    print "Analyzing: $sample\n";
    
    my $INPUT = IO::Uncompress::Gunzip->new($file) or die "Failed to open $file: $GunzipError\n";
    while(<$INPUT>) {
        chomp;
        my @fields = split;
        my ($chr, $pos, $cov, $edit, $lvl) = ($fields[0], $fields[1], $fields[2], $fields[3], $fields[4]);

        next if ($chr eq '#chrom');

        my $site = join '_', $chr, $pos;
        my $ratio = join '/', $edit, $cov;
        if ($cov >= $mincov) {
            $sitehash{$sample}{$site} = $ratio;
            if ($totalhash{$site}) {
                $totalhash{$site}++;
                $lvlhash{$site} = join ',', $lvlhash{$site}, $ratio;
            } else {
                $totalhash{$site} = 1;
                $lvlhash{$site} = $ratio;
            }
        }
    }
    close $INPUT;
}

# Prepare output file
unlink $output_file if -e $output_file;
print "Output file: $output_file\n";
open(my $OUTPUT, ">", $output_file) or die "Could not open file '$output_file': $!\n";

print $OUTPUT "site";
foreach my $sample (@samples) {
	print $OUTPUT "\t$sample";
}
print $OUTPUT "\n";
foreach my $site (keys %totalhash) {
	if ($totalhash{$site} >= $minsamps) {
		my @lvls = split(/\,/, $lvlhash{$site});
		
			print $OUTPUT "$site";
			foreach my $sample (@samples) {
				if ($sitehash{$sample}{$site}) {
					print $OUTPUT "\t$sitehash{$sample}{$site}";
				} else {
					print $OUTPUT "\t0/0";
				}
			}
			print $OUTPUT "\n";
		
	}
}
close $OUTPUT;

File: phenotyping/scripts/RNA_editing/shared_samples_sites_matrix.py
#!/usr/bin/env python3
# Assemble edit level matrix
# Based on sharedsamples_sites_matrix_FastQTL_v8.pl from https://github.com/vargasliqin/GTEx_edQTL

import argparse
import gzip
from pathlib import Path

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(
        description='Create a site-by-sample RNA editing ratio matrix from per-sample editing files'
    )
    parser.add_argument('--path_to_edit_files', required=True,
                        help='Directory containing {sample}.rnaeditlevel.tsv.gz files')
    parser.add_argument('--samples_file', required=True,
                        help='File containing sample IDs, one per line')
    parser.add_argument('--output_file', required=True,
                        help='Output TSV matrix with one row per retained edit site and one column per sample')
    parser.add_argument('--min_coverage', type=int, required=True,
                        help='Minimum read coverage required for a site to count as observed in a sample')
    parser.add_argument('--min_samples', type=int, required=True,
                        help='Minimum number of samples in which a site must meet --min_coverage to be retained')
    args = parser.parse_args()

    # Ensure path ends with slash
    path_edit_files = str(Path(args.path_to_edit_files).resolve()) + '/'

    print(f"Path to Edit Files: {path_edit_files}")
    print(f"Samples File: {args.samples_file}")
    print(f"Output File: {args.output_file}")
    print(f"Min Coverage: {args.min_coverage}")
    print(f"Min Samples: {args.min_samples}")

    # Read samples from file
    with open(args.samples_file, 'r') as f:
        samples = [line.strip() for line in f]

    sitehash = {}  # Dictionary to store sample-site ratios
    totalhash = {}  # Dictionary to store site counts
    lvlhash = {}    # Dictionary to store site levels

    # Process each sample
    for sample in samples:
        file_path = f"{path_edit_files}{sample}.rnaeditlevel.tsv.gz"

        print(f"Analyzing: {sample}")
        
        with gzip.open(file_path, 'rt') as f:
            for line in f:
                fields = line.strip().split()
                if fields[0] == '#chrom':
                    continue

                chr_name, pos, cov, edit, lvl = fields
                site = f"{chr_name}_{pos}"
                ratio = f"{edit}/{cov}"

                if int(cov) >= args.min_coverage:
                    if sample not in sitehash:
                        sitehash[sample] = {}
                    sitehash[sample][site] = ratio

                    if site in totalhash:
                        totalhash[site] += 1
                        lvlhash[site] = f"{lvlhash[site]},{ratio}"
                    else:
                        totalhash[site] = 1
                        lvlhash[site] = ratio

    print(f"Output file: {args.output_file}")

    # Check if any sites meet the minimum sample threshold
    sites_to_write = [site for site in totalhash if totalhash[site] >= args.min_samples]
    if not sites_to_write:
        raise ValueError(f"No sites found with at least {args.min_samples} samples meeting the minimum coverage threshold of {args.min_coverage}")

    # Write output matrix
    with open(args.output_file, 'w') as outfile:
        # Write header
        outfile.write("site")
        for sample in samples:
            outfile.write(f"\t{sample}")
        outfile.write("\n")

        # Write data rows
        for site in sites_to_write:
            outfile.write(site)
            for sample in samples:
                if sample in sitehash and site in sitehash[sample]:
                    outfile.write(f"\t{sitehash[sample][site]}")
                else:
                    outfile.write("\t0/0")
            outfile.write("\n")

if __name__ == "__main__":
    main()
